### DISCUSSION TITLE
double distribution is bimodal and mostly empty

### DISCUSSION THREAD

**Discussion by acthp** [dubzzz/fast-check#4328] [Q&A] [answered]
Just testing the float and double arbs, using sample, like so:

```
var x = fc.double({min: 0, max: 2, noNaN: true});

console.log(fc.sample(x, 1000).join('\n'));
```
The numbers are clustered very, very, very tightly next to the min and max, which wasn't what I was expecting. I was expecting a more or less flat distribution. Is this expected behavior? Including a histogram.

![histo](https://github.com/dubzzz/fast-check/assets/494506/6532b020-6ac8-478e-a365-302193af9962)

Edit to add some additional notes:
I found the "bias" code, which appears to be partially responsible for this behavior. But, this result seems rather extreme, essentially omitting almost the entire range. The integer method is similarly biased, but doesn't omit the entire middle of the range. It's flat in the middle, but is biased to the extremes of the range.

There is a noBias() method which on the integer arb will generate a flat distribution, but on the double arb generates a distribution that is clustered entirely around min.

**Comment by dubzzz** [comment, marked as answer]
The current implementation of `double` is uniform on the set of available values. The reason why it mostly generates values close to zero is that there are far more values in [0,1[ than in [1,2[ for a double.

That's said, if you want a more uniform distribution for the value you might want to rely on something like: `fc.integer({ min: 0, max: (1 << 24) - 1 }).map((v) => v / (1 << 24))` for f32 floats or `fc.tuple(fc.integer({ min: 0, max: (1 << 26) - 1 }), fc.integer({ min: 0, max: (1 << 27) - 1 })).map((v) => (v[0] * Math.pow(2, 27) + v[1]) * Math.pow(2, -53))` for f64 floats.

**Reply by dubzzz** [reply]
@acthp I marked this comment as an answer. Please let me know if it was not a proper answer
